### PR TITLE
INT-2251: Fix linting crash on typescript 4 tuple type by upgrading deps

### PR DESCRIPTION
**Pull request by tiny-james**
Note that I ran into these lints when upgrading on connectfour:
```
        "@typescript-eslint/no-unsafe-assignment": "off",
        "@typescript-eslint/no-unsafe-call": "off",
        "@typescript-eslint/no-unsafe-member-access": "off",
        "@typescript-eslint/no-unsafe-return": "off",
        "@typescript-eslint/explicit-module-boundary-types": "off",
        "@typescript-eslint/restrict-plus-operands": "off"
```
Some nice things there to disallow use of `any` (all the "no-unsafe-x") also the "explicit-module-boundary-types" ensures that functions have their types annotated. I'm dubious about the utility of "restrict-plus-operands" as it will make us use template strings and `toString()` calls all over the place...

### CODE REVIEW

**Comment by lnewson** [comment]
I tried upgrading TinyMCE to use this be I get thousands of errors, so I don't think we can merge this as is. Should we disable those new rules and then add them back in if we want them? or do you think it's better to just disable all rules in the relevant projects? (Note: I haven't actually looked at what the rules do atm)

I should add even after disabling the issues mentioned, I still get a couple hundred errors.

**Comment by lnewson** [comment]
Starting to took through the errors, a few definitely are getting incorrectly flagged, so they should be disabled or tweaked in this repo. eg
```
relative: [ '%' as '%' ],
```
is getting flagged as invalid for the `@typescript-eslint/prefer-as-const` rule.

File: package.json [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
-    "eslint-plugin-prefer-arrow": "^1.1.7",
-    "resolve": "^1.10.0",
-    "typescript": "^3.8.0"
+    "@typescript-eslint/eslint-plugin": "^4.3.0",
+    "@typescript-eslint/parser": "^4.3.0",
+    "eslint": "^7.10.0",
+    "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-prefer-arrow": "^1.2.2",
+    "resolve": "^1.17.0",
+    "typescript": "^4.0.3"
```
Comment (lnewson) [inline package.json:16]: Is this dependency correct? As far as I can see it doesn't require 4.0.3 specifically? Should this actually be a peer or even a dev dependency?

File: package.json [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
-    "@typescript-eslint/eslint-plugin": "^2.29.0",
-    "@typescript-eslint/parser": "^2.29.0",
-    "eslint": "^6.8.0",
-    "eslint-plugin-import": "^2.20.1",
-    "eslint-plugin-prefer-arrow": "^1.1.7",
-    "resolve": "^1.10.0",
-    "typescript": "^3.8.0"
+    "@typescript-eslint/eslint-plugin": "^4.3.0",
+    "@typescript-eslint/parser": "^4.3.0",
+    "eslint": "^7.10.0",
```
Comment (lnewson) [inline package.json:12]: Do we really need 7.10.0, so should this instead just be `^7.10.0`?

File: package.json [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
   "license": "Apache-2.0",
   "dependencies": {
-    "@typescript-eslint/eslint-plugin": "^2.29.0",
-    "@typescript-eslint/parser": "^2.29.0",
-    "eslint": "^6.8.0",
-    "eslint-plugin-import": "^2.20.1",
-    "eslint-plugin-prefer-arrow": "^1.1.7",
-    "resolve": "^1.10.0",
-    "typescript": "^3.8.0"
+    "@typescript-eslint/eslint-plugin": "^4.3.0",
```
Comment (lnewson) [inline package.json:10]: Similarily to eslint, do we actually require something that's in 4.3.0? Remember libraries should specify the minimum version they should need, not just the latest.

**Review by lnewson — COMMENTED** [review]
Not marking as approved at this point, as I just wanted to leave some feedback and see what you think first.

File: package.json (hunk 1 above)
Comment (tiny-james) [reply package.json]: I set it back to ^4

File: package.json (hunk 2 above)
Comment (tiny-james) [reply package.json]: I set it to ^7.0.0

File: package.json (hunk 3 above)
Comment (tiny-james) [reply package.json]: I do not know of any need for 4.3 specifically, it was just the latest available. I have reverted it back to ^4.0.0

**Review by lnewson — APPROVED** [review]

**Comment by tiny-james** [comment]
> I tried upgrading TinyMCE to use this be I get thousands of errors, so I don't think we can merge this as is. Should we disable those new rules and then add them back in if we want them? or do you think it's better to just disable all rules in the relevant projects? (Note: I haven't actually looked at what the rules do atm)
> 
> I should add even after disabling the issues mentioned, I still get a couple hundred errors.

How do you test? I tried using yarn link but it didn't seem to do anything. Do the new disabled rules fix the tinymce issues?

**Comment by lnewson** [comment]
> > I tried upgrading TinyMCE to use this be I get thousands of errors, so I don't think we can merge this as is. Should we disable those new rules and then add them back in if we want them? or do you think it's better to just disable all rules in the relevant projects? (Note: I haven't actually looked at what the rules do atm)
> > I should add even after disabling the issues mentioned, I still get a couple hundred errors.
> 
> How do you test? I tried using yarn link but it didn't seem to do anything. Do the new disabled rules fix the tinymce issues?

I use a tool called `yalc` to test this sort of stuff normally. Normally requires doing something like this:
```
[eslint-plugin](INT-2251)$ yalc publish
[tinymce](develop)$ yalc add @tinymce/eslint-plugin
[tinymce](develop)$ nano .yarnrc // disable the frozen lockfile
[tinymce](develop)$ yarn && yarn eslint
```

As for if it fixed the errors, it didn't disable all the new ones by the looks. Here's a snippet:
```
@ephox/sugar: /home/lnewson/git/tinymce/modules/sugar/src/test/ts/browser/ElementTest.ts
@ephox/sugar:   4:25  warning  'T' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/sugar: ✖ 1 problem (0 errors, 1 warning)
@ephox/agar: /home/lnewson/git/tinymce/modules/agar/src/main/ts/ephox/agar/api/Monitor.ts
@ephox/agar:   7:65  warning  'Monitor' is assigned a value but never used  @typescript-eslint/no-unused-vars
@ephox/agar: /home/lnewson/git/tinymce/modules/agar/src/main/ts/ephox/agar/pipe/Pipe.ts
@ephox/agar:   9:3   error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
@ephox/agar:   9:75  error  Implied eval. Consider passing a function                                                        @typescript-eslint/no-implied-eval
@ephox/agar: /home/lnewson/git/tinymce/modules/agar/src/main/ts/ephox/agar/touch/Touches.ts
@ephox/agar:   19:5  error  Do not use "@ts-ignore" because it alters compilation errors  @typescript-eslint/ban-ts-comment
@ephox/agar: ✖ 4 problems (3 errors, 1 warning)
@ephox/robin: /home/lnewson/git/tinymce/modules/robin/src/test/ts/module/ephox/robin/test/ZoneObjects.ts
@ephox/robin:   26:23  warning  'E' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/robin:   26:26  warning  'D' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/robin: ✖ 2 problems (0 errors, 2 warnings)
@ephox/snooker: /home/lnewson/git/tinymce/modules/snooker/src/test/ts/atomic/calc/DeltasTest.ts
@ephox/snooker:   23:89   error  Invalid type "number[]" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/snooker:   23:110  error  Invalid type "number[]" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/snooker: /home/lnewson/git/tinymce/modules/snooker/src/test/ts/browser/TableAdjustmentsTest.ts
@ephox/snooker:   115:91   error  Invalid type "number[]" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/snooker:   115:124  error  Invalid type "number[]" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/snooker: /home/lnewson/git/tinymce/modules/snooker/src/test/ts/module/ephox/snooker/test/SizeUtils.ts
@ephox/snooker:   51:70  error  Invalid type "string | null" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/snooker: ✖ 5 problems (5 errors, 0 warnings)
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/alien/Descend.ts
@ephox/alloy:   3:35  warning  'T' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/api/behaviour/Keying.ts
@ephox/alloy:   60:18  warning  'S' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/behaviour/docking/DockingState.ts
@ephox/alloy:   12:39  error  Invalid type "boolean" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/alloy:   12:65  error  Invalid type "boolean" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/debugging/FunctionAnnotator.ts
@ephox/alloy:   48:1  error  Expected indentation of 2 spaces but found 4  @typescript-eslint/indent
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/keying/KeyingModeTypes.ts
@ephox/alloy:   35:36  warning  'C' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/test/ts/atomic/api/DomDefinitionTest.ts
@ephox/alloy:   32:24  warning  'T' is defined but never used  @typescript-eslint/no-unused-vars
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/test/ts/browser/behaviour/sliding/SlidingInterruptedTest.ts
@ephox/alloy:   105:76  error  Invalid type "any" of template literal expression  @typescript-eslint/restrict-template-expressions
@ephox/alloy: ✖ 8 problems (4 errors, 4 warnings)
@ephox/alloy:   1 error and 0 warnings potentially fixable with the `--fix` option.
```

**Comment by lnewson** [comment]
```
@ephox/alloy: /home/lnewson/git/tinymce/modules/alloy/src/main/ts/ephox/alloy/alien/Descend.ts
@ephox/alloy:   3:35  warning  'T' is defined but never used  @typescript-eslint/no-unused-vars
```
Is an example of something that is valid though, so not saying we should disable all those rules.

**Comment by tiny-james** [comment]
@lnewson 
I followed your instruction exactly and oddly it claimed there were no linting errors - I am not sure what I'm getting wrong. Anyway I disabled the additional rules you identified in the snippet (the no-unused-vars was already set as a warning so I left it as it was).

**Comment by lnewson** [comment]
> @lnewson
> I followed your instruction exactly and oddly it claimed there were no linting errors - I am not sure what I'm getting wrong. Anyway I disabled the additional rules you identified in the snippet (the no-unused-vars was already set as a warning so I left it as it was).

Thanks! I can confirm I'm only getting the no-unused-var warnings now.

**Review by HAFRMO — APPROVED** [review]

**Review by JamesToohey — APPROVED** [review]

**Pull request merged**